### PR TITLE
[`Awq`] Add llava fused modules support

### PR DESCRIPTION
**Pull request by younesbelkada**
# What does this PR do?

This PR adds the Llava + fused modules support for blazing fast text generation using Llava + AWQ! 

This PR also fixes the issue: https://github.com/huggingface/transformers/pull/28032#issuecomment-1868262141 pointed out by a user since a custom past key value is passed to the model, indeed filtering out indexes that are inside the range of `extended_attention_mask` fixes the issue.

Added also a slow test

Can also confirm all Llava slow tests pass!

cc @casper-hansen

### CODE REVIEW

**Comment by cbjtu** [comment]
Thank you soooo much, this PR and #28032 helped me work well now!

File: src/transformers/modeling_utils.py [hunk 1]
```diff
@@ -3574,6 +3574,14 @@ def from_pretrained(
...
             if quantization_config is None:
                 quantization_config = AwqConfig.from_dict(config.quantization_config)
+            # In case a user passes a `AwqConfig` with `do_fuse=True` for models that have
+            # a `modules_to_not_convert` attribute we need to manually set that attribute into the
+            # passed `quantization_config`
+            elif (
+                quantization_config.modules_to_not_convert is None
+                and "modules_to_not_convert" in config.quantization_config
+            ):
+                quantization_config.modules_to_not_convert = config.quantization_config["modules_to_not_convert"]
```
Comment (ArthurZucker) [inline src/transformers/modeling_utils.py:3584]: that seems a bit odd to me, should either be done in the integration (I know you don't have access to the config) or when you init the `quantization_config`, you should use `config.quantization_config` no? (at some point merging kwargs?)

File: src/transformers/models/llava/modeling_llava.py [hunk 2]
```diff
@@ -445,8 +445,15 @@ def forward(
                         device=attention_mask.device,
                     )
 
+                    # Filter out only the tokens that can be un-attended, this can happen
+                    # in the case one uses Llava + Fused modules where the cache on the
```
Comment (ArthurZucker) [inline src/transformers/models/llava/modeling_llava.py:449]:
```suggestion
                    # if one uses Llava + Fused modules where the cache on the
```

File: src/transformers/models/llava/modeling_llava.py [hunk 3]
```diff
@@ -445,8 +445,15 @@ def forward(
...
+                    # first iteration is already big enough, or if one passes custom cache
+                    valid_indices = non_attended_tokens < extended_attention_mask.size(-1)
+                    new_batch_index = batch_index[valid_indices]
+                    new_non_attended_tokens = non_attended_tokens[valid_indices]
```
Comment (ArthurZucker) [inline src/transformers/models/llava/modeling_llava.py:453]: not a fan of adding custom code only to handle custom usages. There should be a more general way of handling these things (why use the extended attention mask and not just the attention mask, why not use the past key value length, etc)

**Review by ArthurZucker — COMMENTED** [review]
Thanks a lot! More discussion needed for the attended / non attended tokens is needed IMO! 🤗

File: src/transformers/integrations/awq.py [hunk 4]
```diff
@@ -248,7 +265,14 @@ def _fuse_awq_mlp(model, current_module_name, fuse_module_names, module, target_
         down_proj = getattr(module, fuse_module_names[2])
 
         previous_device = gate_proj.qweight.device
-        activation_fn = ACT2FN[model.config.hidden_act]
+
+        # Deal also with the case model has `text_config` attribute
```
[the remark this answers is not in the document]
Comment (younesbelkada) [reply src/transformers/integrations/awq.py:269]: yes awq supports only quantizing text backbones currently

File: src/transformers/modeling_utils.py (hunk 1 above)
Comment (younesbelkada) [reply src/transformers/modeling_utils.py:3598]: I think we always rely on `quantization_config`, we do merge kwargs but on the other way around (from `quantization_config` to `config.quantization_config`) with `get_loading_attributes()`. The scenario above happens only with the specific case where users pass `do_fuse=True` & a non-`None` value in `config.quantization_config["modules_to_not_convert"]`. I think it is a good idea to think of a way to harmonize how to merge kwargs between `config.quantization_config` and `quantization_config` but might be slightly out of the scope of the PR as I need to do it for all quantization schemes we support. I propose to do that properly in a follow up PR

File: src/transformers/models/llava/modeling_llava.py (hunk 3 above)
Comment (younesbelkada) [reply src/transformers/models/llava/modeling_llava.py:461]: This is not to handle custom usage, it happens when a past key value with padd tokens are on indices that are larger than the extended attention mask shape: https://github.com/huggingface/transformers/pull/28032#issuecomment-1868190266 & https://github.com/huggingface/transformers/pull/28239#issuecomment-1869224325 - this can mainly happen in batched generation with long seq len and it specifically happens for autoawq fused modules because the dummy past key values are initialized will all zeros: https://github.com/casper-hansen/AutoAWQ/blob/a3db8099a234a46a21bf5e46340da60da6992e0c/awq/modules/fused/attn.py#L238 
In any case I don't think this will cause any harm since it just filers out indices of padd tokens (that are not attended anyway) that are out of the extended attention mask range, and I confirmed all slow tests pass

File: src/transformers/modeling_utils.py (hunk 1 above)
Comment (ArthurZucker) [reply src/transformers/modeling_utils.py:3598]: Alright let's keep that in mind

File: src/transformers/models/llava/modeling_llava.py (hunk 3 above)
Comment (ArthurZucker) [reply src/transformers/models/llava/modeling_llava.py:461]: Alright, I though this was already solved. No worries them, it's just that tensor indexing might slow things down a bit but is required anyway. I think a refactor might help: 
- Init the embeddings with a different value (like -1 which is might not happen as often as zeros) when we compute the image indexes
- correctly update the attention mask when merging to make sure we keep track of what we computed
I'd be in favor of moving this fix to another PR maybe? WDYT?

**Review by ArthurZucker — APPROVED** [review]
thanks

File: src/transformers/models/llava/modeling_llava.py (hunk 3 above)
Comment (younesbelkada) [reply src/transformers/models/llava/modeling_llava.py:461]: I see thanks ! 

I am happy to explore: 

> Init the embeddings with a different value (like -1 which is might not happen as often as zeros) when we compute the image indexes

In another PR !

> I'd be in favor of moving this fix to another PR maybe? WDYT?

That might be not ideal because if this fix is not introduced, users cannot run llava + fused modules :/ I'll address the points you shared in a follow up PR !

**Comment by younesbelkada** [comment]
Thanks for your reviews @ArthurZucker ! Merging ! I'll address the points you shared in https://github.com/huggingface/transformers/pull/28239#discussion_r1448621388 in another PR as stated in my reply
